src/terminal/parser/stateMachine.cpp

```diff
@@ -1821,6 +1808,118 @@ bool StateMachine::FlushToTerminal()
+    // to fail to optimize the printable-ASCII range check into a subtraction & comparison.
+    return (wch <= 0x1f) | (static_cast<wchar_t>(wch - 0x7f) <= 0x20);
+}
+
+[[msvc::forceinline]] static size_t findActionableFromGroundPlain(const wchar_t* beg, const wchar_t* end, const wchar_t* it) noexcept
+{
+#pragma loop(no_vector)
+    for (; it < end && !isActionableFromGround(*it); ++it)
+    {
+    }
+    return it - beg;
+}
+
+static size_t findActionableFromGround(const wchar_t* data, size_t count) noexcept
+{
+    // The following vectorized code replicates isActionableFromGround which is equivalent to:
+    //   (wch <= 0x1f) || (wch >= 0x7f && wch <= 0x9f)
+    // or rather its more machine friendly equivalent:
+    //   (wch <= 0x1f) | ((wch - 0x7f) <= 0x20)
+#if defined(TIL_SSE_INTRINSICS)
+
+    auto it = data;
+
+    for (const auto end = data + (count & ~size_t{ 7 }); it < end; it += 8)
+    {
+        const auto wch = _mm_loadu_si128(reinterpret_cast<const __m128i*>(it));
+        const auto z = _mm_setzero_si128();
+
+        // Dealing with unsigned numbers in SSE2 is annoying because it has poor support for that.
+        // We'll use subtractions with saturation ("SubS") to work around that. A check like
+        // a < b can be implemented as "max(0, a - b) == 0" and "max(0, a - b)" is what "SubS" is.
+
+        // Check for (wch < 0x20)
+        auto a = _mm_subs_epu16(wch, _mm_set1_epi16(0x1f));
+        // Check for "(wch >= 0x7f && wch <= 0x9f)" by adding 0x10000-0x7f, which overflows to a
+        // negative number if "wch >= 0x7f" and then subtracting 0x9f-0x7f with saturation to an
+        // unsigned number (= can't go lower than 0), which results in all numbers up to 0x9f to be 0.
+        auto b = _mm_subs_epu16(_mm_add_epi32(wch, _mm_set1_epi16(static_cast<short>(0xff81))), _mm_set1_epi16(0x20));
+        a = _mm_cmpeq_epi16(a, z);
+        b = _mm_cmpeq_epi16(b, z);
+
+        const auto c = _mm_or_si128(a, b);
+        const auto mask = _mm_movemask_epi8(c);
+
+        if (mask)
+        {
+            unsigned long offset;
+            _BitScanForward(&offset, mask);
+            it += offset / 2;
+            return it - data;
+        }
+    }
+
+    return findActionableFromGroundPlain(data, data + count, it);
+
+#elif defined(TIL_ARM_NEON_INTRINSICS)
+
+    auto it = data;
+    uint64_t mask;
+
+    for (const auto end = data + (count & ~size_t{ 7 }); it < end;)
+    {
+        const auto wch = vld1q_u16(it);
+        const auto a = vcleq_u16(wch, vdupq_n_u16(0x1f));
+        const auto b = vcleq_u16(vsubq_u16(wch, vdupq_n_u16(0x7f)), vdupq_n_u16(0x20));
+        const auto c = vorrq_u16(a, b);
+
+        mask = vgetq_lane_u64(c, 0);
+        if (mask)
+        {
+            goto exitWithMask;
+        }
+        it += 4;
+
+        mask = vgetq_lane_u64(c, 1);
+        if (mask)
+        {
+            goto exitWithMask;
+        }
+        it += 4;

```

Comment:
`v[c]name[q][_n]_type`
* `v`: vector operation
* `c`: component-wise operation
* `q`: quad double words = 128 bit operation
* `_n`: scalar instead of vector arguments
* name:
  * `ld1`: load bytes (unaligned by default)
  * `dup`: duplicate argument into all components
  * `sub`: subtract
  * `le`: lower or equal
  * `orr`: bitwise or
* type:
  * `s`: signed
  * `u`: unsigned
  * `u8`: 8 bits per component
  * `u16`: 16 bits per component

`vgetq_lane_u64`: extracts the upper/lower 64 bits of the 128 bit vector